src/runtime/future/vanilla_future.c

```diff
@@ -0,0 +1,168 @@
+
+#define perr(m)  // fprintf(stderr, "%s\n", m);
+
+/*
+      The most minified future API possible in encore.
+
+*/
+
+extern bool pony_system_actor(pony_actor_t *actor);      
+static inline void vanilla_future_discharge(pony_ctx_t **ctx, vanilla_future_t * fut);
+static void vanilla_future_finalizer(vanilla_future_t *fut);
+static inline void vanilla_future_gc_send_value(pony_ctx_t *ctx, vanilla_future_t *fut);
+static inline void vanilla_future_gc_recv_value(pony_ctx_t *ctx, vanilla_future_t *fut);
+static inline void vanilla_future_block_actor(pony_ctx_t **ctx, vanilla_future_t *fut);
+extern void encore_future_gc_acquireactor(pony_ctx_t* ctx, pony_actor_t* actor);
+extern void encore_future_gc_acquireobject(pony_ctx_t* ctx, void* p,
+    pony_type_t *t, int mutability);
+
+static void vanilla_future_finalizer(vanilla_future_t *fut)
+{
+  pony_ctx_t* cctx = pony_ctx();
+  vanilla_future_gc_recv_value(cctx, fut);
+  ENC_DTRACE2(FUTURE_DESTROY, (uintptr_t) cctx, (uintptr_t) fut);
+}
+
+pony_type_t *vanilla_future_get_type(vanilla_future_t *fut){
+  return fut->type;
+}
+
+bool vanilla_future_fulfilled(vanilla_future_t *fut)
+{
+  perr("future_fulfilled");
+  bool r;
+  r = __atomic_load_n(&fut->fulfilled, __ATOMIC_SEQ_CST);
+  return r;
+}
+
+static inline void vanilla_future_gc_trace_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  assert(fut);
+  if (fut->type == ENCORE_ACTIVE) {
+    encore_trace_actor(ctx, fut->value.p);
+  } else if (fut->type != ENCORE_PRIMITIVE) {
+    encore_trace_object(ctx, fut->value.p, fut->type->trace);
+  }
+}
+
+static inline void vanilla_future_gc_send_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  pony_gc_send(ctx);
+  vanilla_future_gc_trace_value(ctx, fut);
+  pony_send_done(ctx);
+}
+
+static inline void vanilla_future_gc_recv_value(pony_ctx_t *ctx, vanilla_future_t *fut)
+{
+  pony_gc_recv(ctx);
+  vanilla_future_gc_trace_value(ctx, fut);
+  ponyint_gc_handlestack(ctx);
+}
+
+vanilla_future_t *vanilla_future_mk(pony_ctx_t **ctx, pony_type_t *type)
+{
+  pony_ctx_t *cctx = *ctx;
+  assert(cctx->current);
+  vanilla_future_t *fut = pony_alloc_final(cctx, sizeof(future_t),
+          (void *)&vanilla_future_finalizer);
+  *fut = (vanilla_future_t) { .type = type, .blocking = false };
+  ENC_DTRACE3(FUTURE_CREATE, (uintptr_t) ctx, (uintptr_t) fut, (uintptr_t) type);
+  return fut;
+} 
+
+encore_arg_t vanilla_future_get_actor(pony_ctx_t **ctx, vanilla_future_t *fut)
+{
+  if (!__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    ENC_DTRACE2(FUTURE_BLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+    vanilla_future_block_actor(ctx, fut);
+    ENC_DTRACE2(FUTURE_UNBLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+  }
+  ENC_DTRACE2(FUTURE_GET, (uintptr_t) *ctx, (uintptr_t) fut);
+  return fut->value;
+}
+
+static inline void vanilla_future_block_actor(pony_ctx_t **ctx, vanilla_future_t *fut)
+{
+  perr("future_block_actor");
+  if (__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
```

Comment:
many of these methods seem to be exactly the same.
Would it make sense to have a macro that defines these common methods and the specialised methods in the files?